daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/NormalizeRollbacksSpec.scala

```diff
@@ -0,0 +1,291 @@
+    Top(R(), c1),
+    Top(c1),
+  )
+  test("empty rollback between creates")(
+    Top(c1, R(), c2),
+    Top(c1, c2),
+  )
+  test("sibling empty rollback")(
+    Top(R(), R()),
+    Top(),
+  )
+  test("nested inner empty rollback")(
+    Top(R(c1, R(), c2)),
+    Top(R(c1, c2)),
+  )
+  test("inner empty rollback, within exercise")(
+    Top(E(c1, R(), c2)),
+    Top(E(c1, c2)),
+  )
+  test("nested empty rollback")(
+    Top(R(R())),
+    Top(),
+  )
+  test("nested sibling empty rollback")(
+    Top(R(R(), R())),
+    Top(),
+  )
+
+  // normalization rules #2 or #3
+  test("nested-rollback")(
+    Top(R(R(c1))),
+    Top(R(c1)),
+  )
+
+  // normalization rules #2
+  test("head-nested-rollback")(
+    Top(R(R(c1), c2)),
+    Top(R(c1), R(c2)),
+  )
+  test("head-nested-rollback-cascade")(
+    Top(R(R(R(c1), c2), c3)),
+    Top(R(c1), R(c2), R(c3)),
+  )
+
+  // normalization rules #3
+  test("tail-nested-rollback")(
+    Top(R(c1, R(c2))),
+    Top(R(c1, c2)),
+  )
+  test("tail-nested-rollback-cascade")(
+    Top(R(c1, R(c2, R(c3)))),
+    Top(R(c1, c2, c3)),
+  )
+  test("tail-nested-rollback-cascade-complex")(
+    Top(R(c1, c2, R(c3, R(c4)))),
+    Top(R(c1, c2, c3, c4)),
+  )
+
+  // normalization rules #2 and #3
+  test("mixed-rollback-nesting-A")(
+    Top(R(R(c1), c2, R(c3))),
+    Top(R(c1), R(c2, c3)),
+  )
+  test("mixed-rollback-nesting-B")(
+    Top(R(R(c1), c2, c3, R(c4))),
+    Top(R(c1), R(c2, c3, c4)),
+  )
+
+}
+
+object NormalizeRollbackSpec {
+
+  type Nid = NodeId
+  type Cid = V.ContractId
+  type TX = GenTransaction[Nid, Cid]
+  type Node = GenNode[Nid, Cid]
+  type RB = NodeRollback[Nid]
+
+  def forallNode(tx: TX)(pred: Node => Boolean): Boolean = {
+    tx.fold[Boolean](true) { case (acc, (_, node)) =>
+      acc && pred(node)
+    }
+  }
+
+  def forallRB(tx: TX)(pred: RB => Boolean): Boolean = {
+    forallNode(tx) {
+      case rb: NodeRollback[_] => pred(rb)
+      case _ => true
+    }
+  }
+
+  def isNormalized(tx: TX): Boolean = {
+    tx match {
+      case GenTransaction(nodes, _) =>
+        def isRB(node: Node): Boolean = {
+          node match {
+            case _: NodeRollback[_] => true
+            case _ => false
+          }
+        }
+        def check(rb: RB): Boolean = {
+          val n = rb.children.length
+          (n > 0) && // Normalization rule #1
+          !isRB(nodes(rb.children(0))) && // Normalization rule #2
+          !isRB(nodes(rb.children(n - 1))) // Normalization rule #3
+        }
+        forallRB(tx) { rb =>
+          check(rb)
+        }
+    }
+  }
+
+  // Shape: description of a transaction, with conversions to and from a real tx
+  sealed trait Shape
+  object Shape {
+
+    final case class Top(xs: List[Shape])
+    final case class Create(x: Long) extends Shape
+    final case class Exercise(x: List[Shape]) extends Shape
+    final case class Rollback(x: List[Shape]) extends Shape
+
+    def toTransaction(top: Top): TX = {
```

Comment:
this looks like we can probably simplify the code a fair bit by using the transaction builder.

Reply:
I looked at the transaction builder, and it look really awkward to use in this case.

Reply:
Just a bit curious about what is awkward.

Reply:
Mainly that I am building the transaction bottom up from a declarative description.
But the transaction builder seems to assume that I am building top-down. Create nodes, and then mutably attach children.